tensorflow/core/graph/gpu_fusion_pass.cc

```diff
@@ -0,0 +1,715 @@
+    const Node* actv;
+  };
+
+  bool IsFusionEligible(const Node* n, FusionData* d);
+
+  void CreateFusionOp(const FusionData* d);
+};
+
+//----------------------------------------------------------------------
+
+Status ROCmFusionPass::Run(const GraphOptimizationPassOptions& options) {
+  // enable the fusion pass if the env var TF_ROCM_ENABLE_FUSION is set
+  const char* enable_fusion = getenv("TF_ROCM_ENABLE_FUSION");
+  if (enable_fusion != nullptr) {
+    // Check if the graph is present, should be either in
+    // - options.graph (for all but POST_PARTITIONING grouping)
+    // - options.partition_graphs (for POST_PARTITIONING_grouping)
+    if (options.graph == nullptr && options.partition_graphs == nullptr) {
+      return Status::OK();
+    }
+
+    if (kROCmFusionPassGrouping ==
+        OptimizationPassRegistry::POST_PARTITIONING) {
+      for (auto& pg : *options.partition_graphs) {
+        if (isGpuDevice(pg.first)) {
+          VLOG(kVlogLevel) << "Running ROCmFusionPass on GPU partition : "
+                           << pg.first;
+          // run the pass
+          RunPass(pg.second.get());
+
+        } else {
+          VLOG(kVlogLevel) << "Skipping ROCmFusionPass on non-GPU partition : "
+                           << pg.first;
+        }
+      }
+    } else {
+      VLOG(kVlogLevel) << "Running ROCmFusionPass on entire graph ";
+
+      // run the pass
+      RunPass(options.graph->get());
+    }
+
+  } else {
+    VLOG(kVlogLevel) << "ROCmFusionPass was not enabled!";
+  }
+
+  return Status::OK();
+}
+
+bool ROCmFusionPass::RunPass(Graph* graph) {
+  // DumpGraph("Before running ROCmFusionPass", &**g);
+
+  std::vector<ROCmFusionOpBase*> fusions;
+  std::set<const Node*> fused_nodes;
+
+  // Initialize a vector of all the fusion operations we currently support
+  InitializeFusions(fusions, graph, fused_nodes);
+
+  std::vector<Node*> order;
+  GetPostOrder(*graph, &order);  // This will give us reverse topological sort.
+
+  for (const Node* n : order) {
+    // VLOG(kVlogLevel) << n;
+
+    if (fused_nodes.count(n)) {
+      // We have fused already this node...skip it
+      // Note that we are traversing nodes in reverse topological order, and
+      // matches are found by comparing node sequences from back to front, so
+      // we will hit nodes that we have already fused into a fusion operation.
+      continue;
+    }
+
+    for (auto fusion : fusions) {
+      if (fusion->DoFusion(n)) {
+        // bail out after the first successful fusion ...
+        // cannot do more than one fusions on a op
+        break;
+      }
+    }
+  }
+
+  // Remove fused nodes from the graph
+  RemoveFusedNodes(fused_nodes);
+
+  // DumpGraph("After running ROCmFusionPass", &**g);
+  return true;
+}
+
+void ROCmFusionPass::InitializeFusions(std::vector<ROCmFusionOpBase*>& fusions,
+                                       Graph* g,
+                                       std::set<const Node*>& fused_nodes) {
+  fusions.push_back(
```

Comment:
@deven-amd the life-cycle of `fusions` and `fused_nodes` might need to be reviewed. it seems these 3 instances of `ROCmFusionOpXXX` might be leaking after `fusions` goes out of scope.